### PR TITLE
TS-306 Enable log rotation for diags.log & traffic.out

### PR DESCRIPTION
**Pull request by danobi**
- The commit message contains an architectural overview of this feature
- 'make test' and './traffic_server -R 1' all pass right now
- Manual testing yields correct rotational behavior and unchanged behavior to access & error logs (eg. squid.log, error.log, etc)
- Documentation is coming soon

### CODE REVIEW

File: proxy/shared/Makefile.am [hunk 1]
```diff
@@ -26,6 +26,8 @@ noinst_LIBRARIES = \
 AM_CPPFLAGS = \
   $(iocore_include_dirs) \
   -I$(top_srcdir)/lib/records \
+	-I$(top_srcdir)/lib/ts \
```
Comment (bellofreedom) [inline proxy/shared/Makefile.am]: indent

File: cmd/traffic_cop/traffic_cop.cc [hunk 2]
```diff
@@ -767,6 +767,26 @@ spawn_manager()
...
   }
 
+  // Move any traffic.out that we can not write to, out
+  //  of the way (TSqa2232)
+  // coverity[fs_check_call]
+  if (access(log_file, W_OK) < 0 && errno == EACCES) {
+    char old_log_file[PATH_NAME_MAX];
+    snprintf(old_log_file, sizeof(old_log_file), "%s.old", log_file);
+    // coverity[toctou]
+    rename(log_file, old_log_file);
```
Comment (SolidWallOfCode) [inline cmd/traffic_cop/traffic_cop.cc]: Can the rename fail if there is already a .old version?

File: cmd/traffic_crashlog/traffic_crashlog.cc [hunk 3]
```diff
@@ -91,7 +92,8 @@ main(int /* argc ATS_UNUSED */, const char **argv)
   crashlog_target target;
   pid_t parent = getppid();
 
-  diags = new Diags("" /* tags */, "" /* actions */, stderr);
+  BaseLogFile *base_log_file = new BaseLogFile("stderr");
```
Comment (SolidWallOfCode) [inline cmd/traffic_crashlog/traffic_crashlog.cc]: Why declare `base_log_file`? Just pass `new BaseLogFIle("stderr")` directly to `Diags` constructor.

File: cmd/traffic_crashlog/Makefile.am [hunk 4]
```diff
@@ -20,6 +20,7 @@ bin_PROGRAMS = traffic_crashlog
 AM_CPPFLAGS = \
   $(iocore_include_dirs) \
   -I$(top_srcdir)/lib \
+  -I$(top_srcdir)/lib/ts \
```
Comment (SolidWallOfCode) [inline cmd/traffic_crashlog/Makefile.am]: Why was this added back in? There was a bug specifically to get rid of that (TS-3759).

**Comment by jpeach** [comment]
Can you explain why we need this, rather than adding better support for something like log rotate?

File: iocore/eventsystem/test_Event.cc [hunk 5]
```diff
@@ -91,23 +91,13 @@ reconfigure_diags()
...
-      diags_log_fp = NULL;
-    }
-  }
-
-  diags = new Diags(bdt, bat, diags_log_fp);
+  BaseLogFile *blf = new BaseLogFile(diags_logpath);
+  diags = new Diags(bdt, bat, blf);
 
-  if (diags_log_fp == NULL) {
+  if (blf && blf->m_fp == NULL) {
```
Comment (SolidWallOfCode) [inline iocore/eventsystem/test_Event.cc]: How can `blf` be `NULL` since you just allocated it? And in that case the condition should be `NULL == blf || NULL == blf->m_fp` because if `blf` is NULL you're failed as well.

**Comment by danobi** [comment]
There are a few reasons:
- It makes sense from the user perspective, since they can already control rotation from records.config of the access & error logs
- We have the same functionality already; most of the changes are restructuring things
- The current implementation of logging to traffic.out is dubious; with this we gain more flexibility and control

File: cmd/traffic_crashlog/Makefile.am (hunk 4 above)
Comment (danobi) [inline cmd/traffic_crashlog/Makefile.am]: I'll see if I can change this. Originally I just wanted it to compile.

File: iocore/net/test_certlookup.cc [hunk 6]
```diff
@@ -206,7 +206,8 @@ SSLReleaseContext(SSL_CTX *ctx)
 int
 main(int argc, const char **argv)
 {
-  diags = new Diags(NULL, NULL, stdout);
+  BaseLogFile *blf = new BaseLogFile("stdout");
+  diags = new Diags(NULL, NULL, blf);
```
Comment (SolidWallOfCode) [inline iocore/net/test_certlookup.cc]: I see this pattern a lot. Maybe `Diags` should have a constructor overload that takes a string and does the `BaseLogFile` construction.

File: lib/ts/BaseLogFile.cc [hunk 7]
```diff
@@ -0,0 +1,571 @@
...
+  m_meta_info = NULL;
+
+  log_log_trace("exiting BaseLogFile constructor, m_name=%s, this=%p\n", m_name, this);
+}
+
+/*
+ * This consturctor creates a BaseLogFile based on a given name.
+ * Similar to above constructor, but is overloaded with the object signature
+ */
+BaseLogFile::BaseLogFile(const char *name, uint64_t sig)
```
Comment (SolidWallOfCode) [inline lib/ts/BaseLogFile.cc]: Comments on this and the previous constructor.
Why aren't these all member initializers? E.g. `m_fp = NULL:` instead of `: m_fp(NULL)`?
Also, for constructors this similar it's probably better to only have the 2 argument form with a default value for the second one.

File: lib/ts/BaseLogFile.cc [hunk 8]
```diff
@@ -0,0 +1,571 @@
...
+
+  log_log_trace("exiting BaseLogFile constructor, m_name=%s, this=%p\n", m_name, this);
+}
+
+/*
+ * This copy constructor creates a BaseLogFile based on a given copy.
+ */
+BaseLogFile::BaseLogFile(const BaseLogFile &copy)
+  : m_fp(NULL), m_start_time(copy.m_start_time), m_end_time(0L), m_bytes_written(0), m_name(ats_strdup(copy.m_name)),
+    m_hostname(ats_strdup(m_hostname)), m_is_regfile(false), m_has_signature(copy.m_has_signature), m_signature(copy.m_signature),
```
Comment (SolidWallOfCode) [inline lib/ts/BaseLogFile.cc]: `m_hostname(ats_strdup(m_hostname))` - what is this trying to do?

File: lib/ts/BaseLogFile.cc [hunk 9]
```diff
@@ -0,0 +1,571 @@
...
+BaseLogFile::~BaseLogFile()
+{
+  log_log_trace("entering BaseLogFile destructor, m_name=%s, this=%p\n", m_name, this);
+
+  if (m_is_regfile)
+    close_file();
+  else
+    log_log_trace("not a regular file, not closing, m_name=%s, this=%p\n", m_name, this);
+  if (m_name)
+    ats_free(m_name);
```
Comment (SolidWallOfCode) [inline lib/ts/BaseLogFile.cc]: Did you look at using `ats_scoped_str` for these?

File: cmd/traffic_manager/traffic_manager.cc [hunk 10]
```diff
@@ -565,6 +612,25 @@ main(int argc, const char **argv)
...
+  // we do it so TS is able to create BaseLogFiles for each value
+  if (strcmp(bind_stdout, "") != 0) {
+    lmgmt->proxy_options =
+      (char *)ats_realloc(lmgmt->proxy_options, strlen(lmgmt->proxy_options) + 1 /* space */
+                                                  + strlen("--bind_stdout ") + strlen(bind_stdout) + 1 /* null term */);
+    strcat(lmgmt->proxy_options, " --bind_stdout ");
+    strcat(lmgmt->proxy_options, bind_stdout);
+  }
+
+  if (strcmp(bind_stderr, "") != 0) {
```
Comment (SolidWallOfCode) [inline cmd/traffic_manager/traffic_manager.cc]: Maybe `0 == *bind_stdout` instead of a function call?

File: lib/ts/BaseLogFile.cc [hunk 11]
```diff
@@ -0,0 +1,571 @@
...
+ * disk.
+
+ * If we can't use the meta-file, either because it's not there or because
+ * it's not valid, then we'll use timestamp 0 (Jan 1, 1970) as the starting
+ * bound.
+
+ * Return 1 if file rolled, 0 otherwise
+ */
+int
+BaseLogFile::roll(long interval_start, long interval_end)
```
Comment (SolidWallOfCode) [inline lib/ts/BaseLogFile.cc]: Why `long` vs. `time_t`?

File: lib/ts/BaseLogFile.cc [hunk 12]
```diff
@@ -0,0 +1,571 @@
...
+      start = m_start_time;
+    else
+      start = (m_start_time < interval_start) ? m_start_time : interval_start;
+  }
+  log_log_trace("in BaseLogFile::roll(..), start = %ld, m_start_time = %ld, interval_start = %ld\n", start, m_start_time,
+                interval_start);
+
+  // Now that we have our timestamp values, convert them to the proper
+  // timestamp formats and create the rolled file name.
+  timestamp_to_str((long)start, start_time_ext, 64);
```
Comment (SolidWallOfCode) [inline lib/ts/BaseLogFile.cc]: Don't use raw numbers unless you must. Here use `sizeof(start_time_ext)`.

File: cmd/traffic_cop/traffic_cop.cc (hunk 2 above)
Comment (danobi) [inline cmd/traffic_cop/traffic_cop.cc]: It looks like whether or not rename() fails or overwrites is implementation specific. However, BaseLogFile::open_file() also checks to see if the file can be opened, so at the very least the new code's behavior remains consistent with the current code. Therefore, I think that the potential renaming problem is outside the scope of this issue/patch.

File: lib/ts/BaseLogFile.cc [hunk 13]
```diff
@@ -0,0 +1,571 @@
...
+  // exists.  If so, then we need to add a version tag to the rolled
+  // filename as well so that we don't clobber existing files.
+  int version = 1;
+  while (BaseLogFile::exists(roll_name)) {
+    log_log_trace("The rolled file %s already exists; adding version "
+                  "tag %d to avoid clobbering the existing file.\n",
+                  roll_name, version);
+    snprintf(roll_name, LOGFILE_ROLL_MAXPATHLEN, "%s%s%s.%s-%s.%d%s", m_name, (m_hostname ? LOGFILE_SEPARATOR_STRING : ""),
+             (m_hostname ? m_hostname : ""), start_time_ext, end_time_ext, version, LOGFILE_ROLLED_EXTENSION);
+    version++;
```
Comment (SolidWallOfCode) [inline lib/ts/BaseLogFile.cc]: `++version`. ALways use prefix unless you specifically need postfix.

File: lib/ts/BaseLogFile.cc [hunk 14]
```diff
@@ -0,0 +1,571 @@
...
+ */
+int
+BaseLogFile::open_file(int perm)
+{
+  log_log_trace("BaseLogFile: entered open_file()\n");
+  if (is_open()) {
+    return LOG_FILE_NO_ERROR;
+  }
+
+  if (m_name && !strcmp(m_name, "stdout")) {
```
Comment (SolidWallOfCode) [inline lib/ts/BaseLogFile.cc]: You assume later in this method that `m_name` is not `NULL`. So you should either assume that here as well, or do a pre-check verifying it is not `NULL` and failing out of the method if it is.

File: proxy/logging/LogFile.h [hunk 15]
```diff
@@ -168,8 +85,6 @@ class LogFile : public LogBufferSink, public RefCountObj
 
   static int write_ascii_logbuffer(LogBufferHeader *buffer_header, int fd, const char *path, const char *alt_format = NULL);
   int write_ascii_logbuffer3(LogBufferHeader *buffer_header, const char *alt_format = NULL);
-  static bool rolled_logfile(char *file);
-  static bool exists(const char *pathname);
```
Comment (SolidWallOfCode) [inline proxy/logging/LogFile.h]: Is it worth taking these out? For client convenience it you could leave them in and have them just forward to the `BaseLogFile` methods. Otherwise you're forcing clients to be aware of the implementation details of `LogFile` (e.g. that it uses `BaseLogFile`).

File: lib/ts/BaseLogFile.cc (hunk 8 above)
Comment (danobi) [inline lib/ts/BaseLogFile.cc]: Typo, whoops. Was supposed to be `m_hostname(ats_strdup(copy.m_hostname))`

File: proxy/logging/LogFile.h [hunk 16]
```diff
@@ -177,7 +92,12 @@ class LogFile : public LogBufferSink, public RefCountObj
   off_t
   get_size_bytes() const
   {
-    return m_file_format != LOG_FILE_PIPE ? m_bytes_written : 0;
+    if (m_file_format == LOG_FILE_PIPE)
+      return 0;
+    else if (m_log)
+      return m_log->get_size_bytes();
+    else
+      return 0;
```
Comment (SolidWallOfCode) [inline proxy/logging/LogFile.h]: `return m_file_format == LOG_PIPE_FILE ? 0 : m_log ? m_log->get_size_bytes() : 0;`

Maybe. That's just my personal style.

File: proxy/logging/Makefile.am [hunk 17]
```diff
@@ -27,7 +28,7 @@ AM_CPPFLAGS = \
   -I$(top_srcdir)/proxy/hdrs \
   -I$(top_srcdir)/proxy/shared \
   -I$(top_srcdir)/mgmt \
-  -I$(top_srcdir)/mgmt/utils
+  -I$(top_srcdir)/mgmt/utils
```
Comment (SolidWallOfCode) [inline proxy/logging/Makefile.am]: No trailing spaces.

**Comment by SolidWallOfCode** [comment]
jpeach;

I know this has been discussed quite a bit and I've always heard about doing the rotation for `traffic.out` and `diags.log` (as in the original bug description) and never heard mention of being more supportive of an external rotation tool. I think having two different mechanisms, though, (one for these files and another for the other log files) is the worst choice. We should require an external tool or not for everything.

File: lib/ts/BaseLogFile.cc (hunk 7 above)
Comment (danobi) [inline lib/ts/BaseLogFile.cc]: I would actually like some advice on the constructor arguments part. I originally wanted to make the signature a default parameter, but I couldn't think of a good default value. `BaseLogFile` must be able to support having an optional signature. Unless my understanding of the hash implementation is wrong, theoretically any value that `long` can represent can be a valid hash. That's why I added the `m_has_signature` flag.

File: cmd/traffic_crashlog/Makefile.am (hunk 4 above)
Comment (SolidWallOfCode) [inline cmd/traffic_crashlog/Makefile.am]: Well, Leif will go ballistic if this creeps back in, it was a very painful experience to take it out.

File: lib/ts/BaseLogFile.cc (hunk 11 above)
Comment (danobi) [inline lib/ts/BaseLogFile.cc]: It was originally like that in LogFile.cc. I assume for reasons related to the `snprintf()` calls later.

File: proxy/logging/LogFile.cc [hunk 18]
```diff
@@ -754,115 +585,33 @@ LogFile::display(FILE *fd)
...
-  _filename = (char *)ats_malloc(l + 7);
-
-  if (i < 0) {
-    ink_string_concatenate_strings(_filename, ".", filename, ".meta", NULL);
-  } else {
-    memcpy(_filename, filename, i + 1);
-    ink_string_concatenate_strings(&_filename[i + 1], ".", &filename[i + 1], ".meta", NULL);
-  }
+  if (m_file_format == LOG_FILE_PIPE)
+    return (m_fd >= 0);
```
Comment (SolidWallOfCode) [inline proxy/logging/LogFile.cc]: 'return m_fd >= 0;'

File: lib/ts/BaseLogFile.cc (hunk 7 above)
Comment (SolidWallOfCode) [inline lib/ts/BaseLogFile.cc]: If there's really no good default value to signal not passed, you may have to just have these two. A couple of mitigations -

If you're going to do member initialization via assignment anyway, you could create an `init` method that is called from the constructors to do that thereby not duplicating the code.

More obscurely I have created a super class with the common constructed members and used private or protected inheritance because no client cares about that, but you can then unify the common member initialization there. Some people frown on that, though, because it splits the member declarations from the "real" class. Note that you don't need to qualify any of the member names when used later.

File: lib/ts/BaseLogFile.cc (hunk 9 above)
Comment (danobi) [inline lib/ts/BaseLogFile.cc]: I took a look and (I might be wrong) it looks like `ats_scoped_str` 's only live as long as the most immediate scope. I'm not sure how it could be used for the life of an object.

Comment (SolidWallOfCode) [inline lib/ts/BaseLogFile.cc]: If you declare one as a member, it will have the same life time as the instance of which it is a member.

File: iocore/net/test_certlookup.cc (hunk 6 above)
Comment (danobi) [inline iocore/net/test_certlookup.cc]: If the third parameter is changed into `const char *`, there will exist constructor ambiguities when, for example, `NULL` is the third parameter. I'm probably missing something here, but I think any changes (that I can currently think of) made to the API to disambiguate the parameters would decrease the overall understandability of the API.

File: doc/reference/configuration/records.config.en.rst [hunk 19]
```diff
@@ -214,6 +214,28 @@ A value of ``0`` means no signal will be sent.
...
    processes. If no path is specified, then Traffic Server creates the file in its logging directory.
 
+.. ts:cv:: CONFIG proxy.config.output.logfile.rolling_enabled INT 0
+   :reloadable:
+
+   Specifies how the output log is rolled. You can specify the following values:
+
+   -  ``0`` = disables output log rolling
+   -  ``1`` = enables output log rolling at specific intervals (specified with the
+       `proxy.config.output.logfile.rolling_interval_sec`_ variable). The "clock" starts ticking on Traffic Server boot
```
Comment (SolidWallOfCode) [inline doc/reference/configuration/records.config.en.rst]: ts:cv:`proxy.config.output.logfile.rolling_interval_sec`

File: iocore/eventsystem/test_Buffer.cc [hunk 20]
```diff
@@ -91,23 +91,13 @@ reconfigure_diags()
...
-      diags_log_fp = NULL;
-    }
-  }
-
-  diags = new Diags(bdt, bat, diags_log_fp);
+  BaseLogFile *blf = new BaseLogFile(diags_logpath);
+  diags = new Diags(bdt, bat, blf);
 
-  if (diags_log_fp == NULL) {
+  if (blf && blf->m_fp == NULL) {
```
Comment (SolidWallOfCode) [inline iocore/eventsystem/test_Buffer.cc]: `if (blf == NULL || blf->m_fp == NULL)` - you certainly don't want to continue on if `blf` is `NULL`.

File: lib/ts/BaseLogFile.cc [hunk 21]
```diff
@@ -0,0 +1,578 @@
...
+ * bound.
+
+ * Return 1 if file rolled, 0 otherwise
+ */
+int
+BaseLogFile::roll(long interval_start, long interval_end)
+{
+  // First, let's see if a roll is even needed.
+  if (m_name == NULL || !BaseLogFile::exists((char *)m_name)) {
+    log_log_trace("Roll not needed for %s; file doesn't exist\n", ((char *)m_name) ? (char *)m_name : "no_name\n");
```
Comment (SolidWallOfCode) [inline lib/ts/BaseLogFile.cc]: For scoped strings, you only need to be explicit for macros, not for functions. Most of the time just "m_name" should work. In cases like this, you should use the `get` method instead of casting. E.g.
`log_log_trace("Roll not needed for %s; file doesn't exist\n", m_name ? m_name,get() : "no_name\n")

File: lib/ts/BaseLogFile.cc [hunk 22]
```diff
@@ -0,0 +1,578 @@
...
+  }
+}
+
+/*
+ * Changes names of the actual log file (not metainfo)
+ */
+void
+BaseLogFile::change_name(const char *new_name)
+{
+  ats_free(m_name);
```
Comment (SolidWallOfCode) [inline lib/ts/BaseLogFile.cc]: This will cause a double free.

File: lib/ts/BaseLogFile.cc (hunk 21 above)
Comment (danobi) [inline lib/ts/BaseLogFile.cc]: Without the casting or the `get()` function, I was seeing compiler warnings about non-POD types. I'll change all the casts to `get()` calls instead, unless there's a better way to remove the warnings.

File: lib/ts/Diags.h [hunk 23]
```diff
@@ -36,12 +36,14 @@
 #define __DIAGS_H___
 
 #include <stdarg.h>
-#include "ts/ink_error.h"
-#include "ts/ink_mutex.h"
-#include "ts/Regex.h"
-#include "ts/ink_apidefs.h"
+#include "ink_error.h"
```
Comment (SolidWallOfCode) [inline lib/ts/Diags.h]: Don't make this change. It compiles as it was originally.

File: proxy/Makefile.am [hunk 24]
```diff
@@ -46,7 +46,7 @@ AM_CPPFLAGS = \
   -I$(top_srcdir)/mgmt \
   -I$(top_srcdir)/mgmt/utils \
   -I$(top_srcdir)/proxy/api/ts \
-  -I$(top_srcdir)/lib \
+  -I$(top_srcdir)/lib
```
Comment (SolidWallOfCode) [inline proxy/Makefile.am]: Why was this changed? It looks like it would break @OPENSSL_INCLUDES@. If that's not needed it should be removed entirely. My test indicates it is not needed.

Comment (danobi) [inline proxy/Makefile.am]: Whoops, must not have noticed the `@OPENSSL_INCLUDES@` underneath.

**Comment by SolidWallOfCode** [comment]
That's all my comments for now, looking much better.

**Comment by danobi** [comment]
Thanks for all the input, I appreciate it.
